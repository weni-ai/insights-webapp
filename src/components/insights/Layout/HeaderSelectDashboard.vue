<template>
  <UnnnicDropdown
    class="header-select-dashboard"
    position="bottom-right"
  >
    <template #trigger>
      <UnnnicAvatarIcon
        v-if="isSelectedDefaultDashboard"
        icon="monitoring"
        scheme="aux-purple-500"
        @click.stop
      />
      <UnnnicIcon
        v-else
        class="clickable"
        icon="arrow_back"
        @click.stop="$router.back"
      />
      <section class="dropdown__trigger">
        <h1 class="trigger__title">
          {{
            currentDashboard.name || dashboardDefault.name || dashboards[0].name
          }}
        </h1>
        <UnnnicIcon icon="expand_more" />
      </section>
    </template>
    <UnnnicDropdownItem
      v-for="dashboard of dashboards"
      :key="dashboard"
      @click="setCurrentDashboard(dashboard)"
    >
      {{ dashboard.name }}
    </UnnnicDropdownItem>
  </UnnnicDropdown>
</template>

<script>
import { mapActions, mapGetters, mapState } from 'vuex';

export default {
  name: 'HeaderSelectDashboard',

  computed: {
    ...mapState({
      dashboards: (state) => state.dashboards.dashboards,
      currentDashboard: (state) => state.dashboards.currentDashboard,
    }),
    ...mapGetters({
      dashboardDefault: 'dashboards/dashboardDefault',
    }),

    isSelectedDefaultDashboard() {
      return this.currentDashboard.uuid === this.dashboardDefault.uuid;
    },
  },

  methods: {
    ...mapActions({
      setCurrentDashboard: 'dashboards/setCurrentDashboard',
    }),
  },
};
</script>

<style lang="scss" scoped>
$dropdownFixedWidth: 314px;

.header-select-dashboard {
  display: flex;

  .dropdown__trigger {
    display: flex;
    align-items: center;
    gap: $unnnic-spacing-nano;

    cursor: pointer;

    .trigger__title {
      margin: $unnnic-spacing-nano 0;

      color: $unnnic-color-neutral-darkest;
      font-family: $unnnic-font-family-primary;
      font-size: $unnnic-font-size-title-sm;
      font-weight: $unnnic-font-weight-bold;
      line-height: $unnnic-line-height-large * 2;
    }
  }

  :deep(.unnnic-dropdown__trigger) {
    display: flex;
    gap: $unnnic-spacing-ant;
    align-items: center;

    .unnnic-dropdown__content {
      margin-top: $unnnic-spacing-nano;

      left: 0;

      width: $dropdownFixedWidth;

      padding: $unnnic-spacing-xs;
      gap: $unnnic-spacing-nano;

      .unnnic-dropdown-item {
        color: $unnnic-color-neutral-darkest;
        font-family: $unnnic-font-family-secondary;
        font-size: $unnnic-font-size-body-gt;
        font-weight: $unnnic-font-weight-bold;
        padding: $unnnic-spacing-xs;
        border-radius: $unnnic-border-radius-sm;

        &:hover {
          background-color: $unnnic-color-neutral-lightest;
        }

        &::before {
          content: none;
        }
      }
    }
  }
}
</style>
