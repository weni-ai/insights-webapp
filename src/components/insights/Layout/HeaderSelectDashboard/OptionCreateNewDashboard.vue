<template>
  <UnnnicDropdownItem
    data-testid="option-new-dashboard"
    class="option-create-new-dashboard"
  >
    <UnnnicIcon
      data-testid="option-new-dashboard-icon"
      icon="add"
    />
    {{ $t('insights_header.add_new_dashboard') }}
  </UnnnicDropdownItem>
</template>

<script>
export default {
  name: 'OptionCreateNewDashboard',
};
</script>

<style lang="scss" scoped>
.option-create-new-dashboard {
  padding: $unnnic-spacing-xs;

  display: flex;
  align-items: center;
  justify-content: center;
  gap: $unnnic-spacing-nano;

  border-top: $unnnic-border-width-thinner solid $unnnic-color-neutral-light;

  color: $unnnic-color-neutral-darkest;
  font-family: $unnnic-font-family-secondary;
  font-size: $unnnic-font-size-body-gt;
}
</style>
