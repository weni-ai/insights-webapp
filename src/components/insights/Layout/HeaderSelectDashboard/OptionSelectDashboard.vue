<template>
  <UnnnicDropdownItem
    data-testid="option-select-dashboard"
    class="option-select-dashboard"
    :class="{
      'option-select-dashboard--active':
        currentDashboard.uuid === dashboard.uuid,
    }"
    @click="setCurrentDashboard(dashboard)"
  >
    {{ dashboard.name }}
    <UnnnicIcon
      data-testid="star-icon"
      class="option-select-dashboard__star-icon"
      :class="{
        'option-select-dashboard__star-icon--selected': isDefaultDashboard,
      }"
      icon="star_rate"
      scheme="neutral-clean"
      clickable
      :filled="isDefaultDashboard || starHovered"
      @mouseenter="setStarHovered(true)"
      @mouseleave="setStarHovered(false)"
      @click.stop="handleSetDefaultDashboard"
    />
  </UnnnicDropdownItem>
</template>

<script>
import { mapActions, mapGetters, mapState } from 'vuex';
import Unnnic from '@weni/unnnic-system';

export default {
  name: 'OptionSelectDashboard',

  props: {
    dashboard: {
      type: Object,
      default: () => {},
    },
  },

  data() {
    return {
      dashboardHovered: '',
      starHovered: false,
    };
  },

  computed: {
    ...mapState({
      currentDashboard: (state) => state.dashboards.currentDashboard,
    }),
    ...mapGetters({
      dashboardDefault: 'dashboards/dashboardDefault',
    }),

    isDefaultDashboard() {
      return this.dashboardDefault?.uuid === this.dashboard.uuid;
    },
  },

  methods: {
    ...mapActions({
      setCurrentDashboard: 'dashboards/setCurrentDashboard',
      setDefaultDashboard: 'dashboards/setDefaultDashboard',
    }),

    setStarHovered(boolean) {
      this.starHovered = boolean;
    },

    async handleSetDefaultDashboard() {
      const { dashboard } = this;
      if (dashboard.uuid === this.dashboardDefault.uuid) return;

      try {
        await this.setDefaultDashboard(dashboard.uuid);
        this.callSetDashboardAlert('success');
      } catch (error) {
        console.error(error);
        this.callSetDashboardAlert('error');
      }
    },

    callSetDashboardAlert(type) {
      const { dashboard } = this;

      if (!['success', 'error'].includes(type)) {
        throw new Error(
          'Error calling the alert when setting the default dashboard. This type does not exist.',
        );
      }

      Unnnic.unnnicCallAlert({
        props: {
          text: this.$t(`insights_header.set_default_dashboard_${type}`, {
            dashboard: dashboard.name,
          }),
          type,
        },
        seconds: 5,
      });
    },
  },
};
</script>

<style lang="scss" scoped>
.option-select-dashboard {
  &.unnnic-dropdown-item::before {
    content: none;
  }

  border-radius: $unnnic-border-radius-sm;

  padding: $unnnic-spacing-xs;

  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: $unnnic-spacing-nano;

  color: $unnnic-color-neutral-darkest;
  font-family: $unnnic-font-family-secondary;
  font-size: $unnnic-font-size-body-gt;

  &--active {
    background-color: $unnnic-color-neutral-lightest;
    font-weight: $unnnic-font-weight-bold;
  }

  .option-select-dashboard__star-icon:not(
      .option-select-dashboard__star-icon--selected
    ):hover {
    color: $unnnic-color-weni-500;
  }

  .option-select-dashboard__star-icon--selected {
    color: $unnnic-color-weni-600;
  }
}
</style>
