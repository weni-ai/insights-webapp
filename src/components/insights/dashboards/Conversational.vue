<template>
  <section class="dashboard-conversational">
    <DashboardHeader class="dashboard-conversational__header" />

    <MostTalkedAboutTopicsWidget
      class="dashboard-conversational__most-talked-about-topics"
    />

    <ConversationalDynamicWidget
      v-for="(widget, index) in orderedDynamicWidgets"
      :key="index"
      :type="widget.type"
      :uuid="widget.uuid"
      class="dashboard-conversational__dynamic-widget"
      :class="{
        'dashboard-conversational__dynamic-widget--only-add': isOnlyAddWidget(
          widget.type,
        ),
      }"
    />

    <CustomizableDrawer />
  </section>
</template>

<script setup lang="ts">
import { computed, onUnmounted, ref, watch } from 'vue';
import { storeToRefs } from 'pinia';
import DashboardHeader from '@/components/insights/conversations/DashboardHeader.vue';
import MostTalkedAboutTopicsWidget from '@/components/insights/conversations/MostTalkedAboutTopicsWidget/index.vue';
import ConversationalDynamicWidget from '@/components/insights/conversations/ConversationalDynamicWidget.vue';
import { useConversationalWidgets } from '@/store/modules/conversational/widgets';
import { useWidgets } from '@/store/modules/widgets';
import CustomizableDrawer from '@/components/insights/conversations/CustomizableWidget/CustomizableDrawer.vue';
import { useCustomWidgets } from '@/store/modules/conversational/customWidgets';

type ConversationalWidgetType = 'csat' | 'nps' | 'add' | 'custom';

const customWidgets = useCustomWidgets();
const { getCustomWidgets } = storeToRefs(customWidgets);

const conversationalWidgets = useConversationalWidgets();
const { isCsatConfigured, isNpsConfigured, getDynamicWidgets } = storeToRefs(
  conversationalWidgets,
);
const widgets = useWidgets();
const { isLoadingCurrentDashboardWidgets, currentDashboardWidgets } =
  storeToRefs(widgets);

const dynamicWidgets = ref<{ type: ConversationalWidgetType; uuid: string }[]>(
  [],
);

const orderedDynamicWidgets = computed(() => {
  if (isLoadingCurrentDashboardWidgets.value) {
    return [];
  }

  return dynamicWidgets.value;
});

const isOnlyAddWidget = (widget: ConversationalWidgetType) => {
  const isOddNumberOfWidgets = dynamicWidgets.value.length % 2 === 1;
  const isLastWidgetAdd = widget === 'add';

  return isOddNumberOfWidgets && isLastWidgetAdd;
};

watch(currentDashboardWidgets, () => {
  if (isCsatConfigured.value) {
    dynamicWidgets.value.push({ type: 'csat', uuid: '' });
  }

  if (isNpsConfigured.value) {
    dynamicWidgets.value.push({ type: 'nps', uuid: '' });
  }

  if (getCustomWidgets.value.length > 0) {
    getCustomWidgets.value.forEach((widget) => {
      dynamicWidgets.value.push({ type: 'custom', uuid: widget.uuid });
    });
  }

  dynamicWidgets.value.push({ type: 'add', uuid: '' });
});

onUnmounted(() => {
  dynamicWidgets.value = [];
});
</script>

<style lang="scss" scoped>
$layout-gap: $unnnic-spacing-sm;

.dashboard-conversational {
  display: flex;
  gap: $layout-gap;
  flex-wrap: wrap;

  &__header {
    width: 100%;
  }

  &__most-talked-about-topics {
    width: 100%;
  }

  &__dynamic-widget {
    min-width: calc(50% - $layout-gap / 2);
    width: fit-content;

    &--only-add {
      min-width: 100%;
    }
  }
}
</style>
