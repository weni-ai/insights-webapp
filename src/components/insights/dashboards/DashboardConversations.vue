<template>
  <section class="dashboard-conversations">
    <TotalConversationsWidget />
    <section class="dashboard-conversations__widget-issues-container">
      <TopicsWidget class="dashboard-conversations__topics-widget" />
      <QueueServiceWidget
        class="dashboard-conversations__queue-service-widget"
      />
    </section>
  </section>
  <section v-if="isLoading">Loading...</section>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import TotalConversationsWidget from '@/components/insights/widgets/Conversations/TotalConversationsWidget.vue';
import QueueServiceWidget from '@/components/insights/widgets/Conversations/QueueServiceWidget.vue';
import TopicsWidget from '@/components/insights/widgets/Conversations/TopicsWidget.vue';

const isLoading = ref(false);
</script>

<style lang="scss" scoped>
.dashboard-conversations {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-sm;

  &__widget-issues-container {
    display: flex;
    gap: $unnnic-spacing-md;
    width: 100%;
    height: 100%;
  }

  &__topics-widget {
    flex: 6;
  }

  &__queue-service-widget {
    flex: 4;
  }
}
</style>
