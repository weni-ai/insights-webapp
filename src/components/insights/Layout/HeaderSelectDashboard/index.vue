<template>
  <UnnnicDropdown
    class="header-select-dashboard"
    position="bottom-right"
    data-onboarding-id="select-dashboard"
  >
    <template #trigger>
      <UnnnicAvatarIcon
        v-if="$route.name === 'dashboard'"
        data-testid="dashboard-icon"
        icon="monitoring"
        scheme="aux-purple-500"
      />
      <UnnnicIcon
        v-else
        data-testid="back-icon"
        class="header-select-dashboard__arrow-back"
        icon="arrow_back"
        scheme="neutral-darkest"
        clickable
        @click.stop="$router.back"
      />
      <section
        data-testid="dropdown-trigger"
        class="dropdown__trigger"
        @mouseup="callTourNextStep('dashboard-onboarding-tour')"
      >
        <h1
          data-testid="dashboard-title"
          class="trigger__title"
        >
          {{ dashboardTitle }}
        </h1>
        <UnnnicIcon
          data-testid="expand-icon"
          icon="expand_more"
        />
      </section>
    </template>

    <OptionSelectDashboard
      v-for="dashboard of dashboards"
      :key="dashboard"
      data-testid="select-dashboard-item"
      :dashboard="dashboard"
    />

    <OptionCreateNewDashboard
      v-if="enableCreateCustomDashboards"
      data-testid="add-new-dashboard-button"
      data-onboarding-id="create-dashboard-button"
      @click="handlerCreateDashboardClick()"
    />
  </UnnnicDropdown>
  <DrawerDashboardConfig
    v-if="showDashboardConfig"
    v-model="showDashboardConfig"
    data-testid="drawer-dashboard-config"
    @close="setShowDashboardConfig(false)"
  />
</template>

<script>
import { mapActions, mapGetters, mapMutations, mapState } from 'vuex';

import OptionSelectDashboard from './OptionSelectDashboard.vue';
import OptionCreateNewDashboard from './OptionCreateNewDashboard.vue';
import DrawerDashboardConfig from '@/components/insights/dashboards/DrawerDashboardConfig.vue';

export default {
  name: 'HeaderSelectDashboard',

  components: {
    OptionSelectDashboard,
    OptionCreateNewDashboard,
    DrawerDashboardConfig,
  },

  computed: {
    ...mapState({
      dashboards: (state) => state.dashboards.dashboards,
      currentDashboard: (state) => state.dashboards.currentDashboard,
      enableCreateCustomDashboards: (state) =>
        state.config.enableCreateCustomDashboards,
      showCreateDashboardTour: (state) =>
        state.onboarding.showCreateDashboardOnboarding,
      onboardingRefs: (state) => state.onboarding.onboardingRefs,
      showDashboardConfig: (state) => state.dashboards.showDashboardConfig,
    }),
    ...mapGetters({
      dashboardDefault: 'dashboards/dashboardDefault',
    }),
    dashboardTitle() {
      const title =
        this.currentDashboard.name ||
        this.dashboardDefault.name ||
        this.dashboards[0].name ||
        '';
      return this.$t(title);
    },
  },
  mounted() {
    this.$nextTick(() => {
      this.setOnboardingRef({
        key: 'select-dashboard',
        ref: document.querySelector('[data-onboarding-id="select-dashboard"]'),
      });
    });
  },
  methods: {
    ...mapMutations({
      setOnboardingRef: 'onboarding/SET_ONBOARDING_REF',
      setShowDashboardConfig: 'dashboards/SET_SHOW_DASHBOARD_CONFIG',
    }),

    ...mapActions({
      beforeOpenDashboardList: 'onboarding/beforeOpenDashboardList',
      callTourNextStep: 'onboarding/callTourNextStep',
    }),

    handlerCreateDashboardClick() {
      this.setShowDashboardConfig(true);
      this.callTourNextStep('dashboard-onboarding-tour');
    },
  },
};
</script>

<style lang="scss" scoped>
$dropdownFixedWidth: 314px;

.header-select-dashboard {
  display: flex;

  .dropdown__trigger {
    display: flex;
    align-items: center;
    gap: $unnnic-spacing-nano;

    cursor: pointer;

    .trigger__title {
      margin: $unnnic-spacing-nano 0;

      color: $unnnic-color-neutral-darkest;
      font-family: $unnnic-font-family-primary;
      font-size: $unnnic-font-size-title-sm;
      font-weight: $unnnic-font-weight-bold;
      line-height: $unnnic-line-height-large * 2;
    }
  }

  &__arrow-back {
    margin: $unnnic-spacing-xs;
  }

  :deep(.unnnic-dropdown__trigger) {
    display: flex;
    gap: $unnnic-spacing-ant;
    align-items: center;

    .unnnic-dropdown__content {
      margin-top: $unnnic-spacing-nano;

      left: 0;

      width: $dropdownFixedWidth;

      padding: $unnnic-spacing-xs;
      gap: $unnnic-spacing-nano;
    }
  }
}
</style>
