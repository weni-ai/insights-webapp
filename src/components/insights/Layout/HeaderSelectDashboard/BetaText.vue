<template>
  <section
    v-if="isConversationalDashboard || isHumanSupportDashboard"
    class="beta_text_container"
    data-testid="beta-container"
  >
    <p
      class="beta_text"
      data-testid="beta-text"
    >
      BETA
    </p>
  </section>
</template>

<script setup lang="ts">
import { useDashboards } from '@/store/modules/dashboards';
import { storeToRefs } from 'pinia';
import { computed } from 'vue';

const { currentDashboard } = storeToRefs(useDashboards());

const isConversationalDashboard = computed(() => {
  return currentDashboard.value.name === 'conversations_dashboard.title';
});

const isHumanSupportDashboard = computed(() => {
  return currentDashboard.value.name === 'human_support_dashboard.title';
});
</script>

<style lang="scss" scoped>
.beta_text_container {
  display: flex;
  padding: 0 $unnnic-spacing-nano;
  justify-content: center;
  align-items: center;
  gap: 10px;

  border-radius: $unnnic-border-radius-sm;
  background: $unnnic-color-weni-50;
}
.beta_text {
  color: $unnnic-color-weni-600;
  font-family: $unnnic-font-family-secondary;
  font-size: $unnnic-font-size-body-md;
  font-weight: $unnnic-font-weight-bold;
  line-height: $unnnic-line-height-md + $unnnic-font-size-body-md;
}
</style>
