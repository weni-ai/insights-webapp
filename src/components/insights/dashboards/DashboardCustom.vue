<template>
  <section
    class="dashboard"
    :style="dashboardGridStyle"
  >
    <WidgetOnboarding
      v-if="showConfigWidgetOnboarding"
      :showCardTour="showOnboarding.card"
      :showWidgetTour="showOnboarding.empty_widget"
    />
    <section
      v-if="isLoadingCurrentDashboardWidgets"
      class="dashboard__loading"
    >
      <IconLoading />
    </section>
    <template
      v-for="widget of currentDashboardWidgets"
      v-else
      :key="widget.uuid"
    >
      <DynamicWidget
        :style="getWidgetStyle(widget.grid_position)"
        :widget="widget"
        :data-onboarding-id="getWidgetOnboardingId(widget)"
        @open-config="handlerWidgetOpenConfig(widget)"
        @click-data="openFlowResultContactList"
      />
    </template>

    <DrawerConfigGallery
      v-if="!!currentWidgetEditing"
      :modelValue="!!currentWidgetEditing"
      @close="updateCurrentWidgetEditing(null)"
    />
    <FlowResultContactListModal
      v-if="showFlowResultsContactListModal"
      :flowResultLabel="flowResultsContactListParams?.label"
      :flow="flowResultsContactListParams?.flow"
      @close="closeFlowResultContactList()"
    />
  </section>
</template>

<script>
import { defineAsyncComponent } from 'vue';
import { mapActions, mapState } from 'pinia';

import { useDashboards } from '@/store/modules/dashboards';
import { useWidgets } from '@/store/modules/widgets';
import { useOnboarding } from '@/store/modules/onboarding';

import DrawerConfigGallery from '@/components/insights/drawers/DrawerConfigGallery/index.vue';
import IconLoading from '@/components/IconLoading.vue';
import WidgetOnboarding from '@/components/insights/onboardings/WidgetOnboarding.vue';
import FlowResultContactListModal from '@/components/FlowResultContactListModal.vue';

export default {
  name: 'DashboardCustom',

  components: {
    DynamicWidget: defineAsyncComponent(
      () => import('@/components/insights/widgets/DynamicWidget.vue'),
    ),
    DrawerConfigGallery,
    IconLoading,
    WidgetOnboarding,
    FlowResultContactListModal,
  },

  data() {
    return {
      showDrawerConfigWidget: false,
      widgetConfigurating: null,
      showOnboarding: {
        card: false,
        empty_widget: false,
      },
      showFlowResultsContactListModal: false,
      flowResultsContactListParams: null,
    };
  },

  computed: {
    ...mapState(useDashboards, ['currentDashboard']),
    ...mapState(useWidgets, [
      'currentDashboardWidgets',
      'currentWidgetEditing',
      'isLoadingCurrentDashboardWidgets',
    ]),
    ...mapState(useOnboarding, ['showConfigWidgetOnboarding']),

    isCustomDashboard() {
      return this.currentDashboard.is_deletable;
    },

    dashboardGridStyle() {
      const { grid } = this.currentDashboard || {};
      if (grid) {
        return {
          gridTemplateColumns: `repeat(${grid.columns}, 1fr)`,
          gridTemplateRows: `repeat(${grid.rows}, 1fr)`,
        };
      }
      return {};
    },
  },

  watch: {
    currentDashboardWidgets: {
      immediate: true,
      handler(newCurrentDashboardWidgets) {
        if (
          !!newCurrentDashboardWidgets &&
          this.isCustomDashboard &&
          !this.isLoadingCurrentDashboardWidgets
        ) {
          this.handlerWidgetsOnboarding();
        }
      },
    },
  },

  methods: {
    ...mapActions(useWidgets, ['updateCurrentWidgetEditing']),
    ...mapActions(useOnboarding, [
      'callTourNextStep',
      'setShowConfigWidgetsOnboarding',
    ]),

    openFlowResultContactList(data) {
      this.flowResultsContactListParams = data;
      this.showFlowResultsContactListModal = true;
    },

    closeFlowResultContactList() {
      this.flowResultsContactListParams = {};
      this.showFlowResultsContactListModal = false;
    },

    handleWidgetFilledData() {
      const hasCard = this.currentDashboardWidgets.filter(
        (e) => e.type === 'card',
      );

      const existFunnel = this.currentDashboardWidgets.some(
        (e) => e.type === 'graph_funnel',
      );
      const existVtex = this.currentDashboardWidgets.some(
        (e) => e.type === 'vtex_order',
      );

      this.showOnboarding = {
        card:
          hasCard.length > 0
            ? !!hasCard.every((widget) => widget.name === '')
            : false,
        empty_widget:
          !!this.currentDashboardWidgets.some(
            (widget) => widget.type === 'empty_column',
          ) &&
          !existFunnel &&
          !existVtex,
      };
    },

    handlerWidgetsOnboarding() {
      const hasWidgetsOnboardingComplete =
        localStorage.getItem('hasWidgetsOnboardingComplete') === 'true';

      if (!hasWidgetsOnboardingComplete) {
        this.handleWidgetFilledData();

        if (!this.showOnboarding.card && !this.showOnboarding.empty_widget) {
          localStorage.setItem('hasWidgetsOnboardingComplete', 'true');
        }

        if (this.showOnboarding.card || this.showOnboarding.empty_widget) {
          this.setShowConfigWidgetsOnboarding(true);
        }
      }
    },

    handlerWidgetOpenConfig(widget) {
      const isNewWidget = this.currentWidgetEditing?.uuid !== widget.uuid;
      if (isNewWidget) {
        this.updateCurrentWidgetEditing(widget).then(() => {
          this.callTourNextStep('widgets-onboarding-tour');
        });
      }
    },

    getWidgetStyle(gridPosition) {
      return {
        gridColumn: `${gridPosition.column_start} / ${gridPosition.column_end + 1}`,
        gridRow: `${gridPosition.row_start} / ${gridPosition.row_end + 1}`,
      };
    },

    getWidgetOnboardingId(widget) {
      return widget.type === 'card'
        ? 'widget-card-metric'
        : 'widget-graph-empty';
    },
  },
};
</script>

<style lang="scss" scoped>
.dashboard {
  overflow: hidden;

  height: 100%;

  display: grid;
  gap: $unnnic-spacing-sm;
  &__loading {
    width: 100vw;
    height: 85vh;
    display: flex;
    justify-content: center;
    align-items: center;
  }
}
</style>
