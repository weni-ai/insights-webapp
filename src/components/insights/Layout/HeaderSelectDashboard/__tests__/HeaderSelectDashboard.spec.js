import { beforeEach, describe, expect, it, vi } from 'vitest';
import { flushPromises, mount } from '@vue/test-utils';
import { createRouter, createWebHistory } from 'vue-router';
import { createI18n } from 'vue-i18n';

import { createTestingPinia } from '@pinia/testing';
import { routes } from '@/router';

import HeaderSelectDashboard from '@/components/insights/Layout/HeaderSelectDashboard/index.vue';
import { useDashboards } from '@/store/modules/dashboards';
import { useConfig } from '@/store/modules/config';

describe('HeaderSelectDashboard', () => {
  let wrapper;
  const router = createRouter({
    history: createWebHistory(),
    routes,
  });

  const store = createTestingPinia({
    initialState: {
      dashboards: {
        dashboards: [
          { name: 'Dashboard 1', uuid: '1' },
          { name: 'Dashboard 2', uuid: '2' },
        ],
        currentDashboard: { name: 'Dashboard 1', uuid: '1' },
        showDashboardConfig: false,
      },
      config: { enableCreateCustomDashboards: false },
      onboarding: {
        onboardingRefs: {
          'select-dashboard': null,
          'create-dashboard-button': null,
          'widget-card-metric': null,
          'widget-gallery': null,
          'drawer-card-metric-config': null,
          'widget-graph-empty': null,
          'drawer-graph-empty': null,
          'dashboard-onboarding-tour': null,
          'widgets-onboarding-tour': null,
        },
        showCreateDashboardOnboarding: false,
        showConfigWidgetOnboarding: false,
        showCompleteOnboardingModal: false,
      },
    },
  });

  const createWrapper = (props) => {
    const i18n = createI18n({
      legacy: false,
      locale: 'en',
      messages: {
        en: {
          'Dashboard 1': 'Dashboard 1',
          'Dashboard 2': 'Dashboard 2',
        },
      },
    });

    return mount(HeaderSelectDashboard, {
      props,
      global: {
        plugins: [router, store, i18n],
        stubs: {
          UnnnicAvatarIcon: true,
          UnnnicIcon: true,
          UnnnicDropdown: {
            template: '<div><slot name="trigger" /><slot /></div>',
          },
          OptionSelectDashboard: true,
          OptionCreateNewDashboard: true,
          DrawerDashboardConfig: true,
          BetaText: true,
        },
      },
    });
  };

  beforeEach(async () => {
    router.push('/');
    wrapper = createWrapper({});
    await flushPromises();
  });

  describe('Title icon', () => {
    it('Should display a "monitoring" icon when the current route is named "dashboard"', async () => {
      router.push({ name: 'dashboard', params: { dashboardUuid: '1' } });
      await flushPromises();

      const dashboardIcon = wrapper.findComponent(
        '[data-testid=dashboard-icon]',
      );
      expect(dashboardIcon.exists()).toBe(true);
      expect(dashboardIcon.props('icon')).toBe('monitoring');
    });

    it('Should display an "arrow_back" icon when the current route is not "dashboard"', async () => {
      router.push({
        name: 'report',
        params: { dashboardUuid: '1', widgetUuid: '2' },
      });
      await flushPromises();

      const dashboardIcon = wrapper.findComponent(
        '[data-testid=dashboard-icon]',
      );
      const backIcon = wrapper.findComponent('[data-testid=back-icon]');

      expect(dashboardIcon.exists()).toBe(false);
      expect(backIcon.exists()).toBe(true);
      expect(backIcon.props('icon')).toBe('arrow_back');
    });

    it('Should redirect to the previous page when the "arrow_back" icon is clicked', async () => {
      vi.spyOn(router, 'back');

      router.push({
        name: 'report',
        params: { dashboardUuid: '1', widgetUuid: '2' },
      });
      await flushPromises();

      const backIcon = wrapper.findComponent('[data-testid=back-icon]');
      await backIcon.trigger('click');

      expect(router.back).toHaveBeenCalled();
    });
  });

  describe('Title', () => {
    it('Should display the name of the currentDashboard if available', () => {
      const dashboardTitle = wrapper.find('[data-testid=dashboard-title]');

      expect(dashboardTitle.exists()).toBe(true);
      expect(dashboardTitle.text()).toBe('Dashboard 1');
    });
  });

  describe('Dropdown behavior', () => {
    it('Should display an "expand_more" icon', () => {
      const expandIcon = wrapper.findComponent('[data-testid=expand-icon]');

      expect(expandIcon.exists()).toBe(true);
      expect(expandIcon.props('icon')).toBe('expand_more');
    });

    it('Should have a dropdown with links matching the length of the dashboards array', async () => {
      const dropdownTrigger = wrapper.find('[data-testid=dropdown-trigger]');
      await dropdownTrigger.trigger('click');
      const selectDashboardItems = wrapper.findAll(
        '[data-testid=select-dashboard-item]',
      );

      const dashboardsStore = useDashboards();

      expect(selectDashboardItems.length).toBe(
        dashboardsStore.dashboards.length,
      );
    });
  });

  describe('Add new dashboard', () => {
    let addNewDashboardButton;

    beforeEach(async () => {
      const configStore = useConfig();
      configStore.enableCreateCustomDashboards = true;

      const dropdownTrigger = wrapper.find('[data-testid=dropdown-trigger]');
      await dropdownTrigger.trigger('click');

      addNewDashboardButton = wrapper.findComponent(
        '[data-testid=add-new-dashboard-button]',
      );
    });

    it('Should display "Create New Dashboard" option when custom dashboards are enabled in the project', async () => {
      expect(addNewDashboardButton.exists()).toBe(true);

      const configStore = useConfig();
      configStore.enableCreateCustomDashboards = false;

      await wrapper.vm.$nextTick();
      expect(addNewDashboardButton.exists()).toBe(false);
    });

    it('Should open the DrawerDashboardConfig when the "Create New Dashboard" option is clicked', async () => {
      const dashboardsStore = useDashboards();
      dashboardsStore.setShowDashboardConfig = (payload) =>
        (dashboardsStore.showDashboardConfig = payload);

      const drawerDashboardConfigBeforeClick = wrapper.findComponent(
        '[data-testid=drawer-dashboard-config]',
      );

      expect(drawerDashboardConfigBeforeClick.exists()).toBe(false);

      await addNewDashboardButton.trigger('click');

      await wrapper.vm.$nextTick();

      const drawerDashboardConfigAfterClick = wrapper.findComponent(
        '[data-testid=drawer-dashboard-config]',
      );
      expect(drawerDashboardConfigAfterClick.exists()).toBe(true);
    });
  });

  it('Should match the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
