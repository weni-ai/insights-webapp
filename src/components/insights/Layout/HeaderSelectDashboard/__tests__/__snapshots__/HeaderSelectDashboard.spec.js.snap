// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeaderSelectDashboard > Should match the snapshot 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="unnnic-dropdown header-select-dashboard"
    data-onboarding-id="select-dashboard"
    data-v-054061ac=""
    data-v-36eb394f=""
  >
    <div
      class="unnnic-dropdown__trigger"
      data-v-36eb394f=""
    >
      
      <span
        class="material-symbols-rounded unnnic-icon-scheme--neutral-darkest unnnic-icon-size--md unnnic--clickable header-select-dashboard__arrow-back"
        data-testid="back-icon"
        data-v-054061ac=""
        data-v-9846e6db=""
        translate="no"
      >
        arrow_back
      </span>
      <section
        class="dropdown__trigger"
        data-testid="dropdown-trigger"
        data-v-054061ac=""
      >
        <h1
          class="trigger__title"
          data-testid="dashboard-title"
          data-v-054061ac=""
        >
          Dashboard 1
        </h1>
        <span
          class="material-symbols-rounded unnnic-icon-scheme--neutral-darkest unnnic-icon-size--md"
          data-testid="expand-icon"
          data-v-054061ac=""
          data-v-9846e6db=""
          translate="no"
        >
          expand_more
        </span>
      </section>
      
      <!---->
    </div>
  </div>
  <drawer-dashboard-config-stub
    data-testid="drawer-dashboard-config"
    data-v-054061ac=""
    modelvalue="true"
  />
  
</div>
`;
